cpp/include/cudf/detail/indexalator.cuh

```diff
@@ -0,0 +1,454 @@
+    return tmp;
+  }
+
+  /**
+   * @brief Compound assignment by sum operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE T& operator+=(difference_type offset)
+  {
+    T& derived = static_cast<T&>(*this);
+    derived.p_ += offset * width_;
+    return derived;
+  }
+
+  /**
+   * @brief Increment by offset operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE T operator+(difference_type offset) const
+  {
+    auto tmp = T(static_cast<T const&>(*this));
+    tmp.p_ += (offset * width_);
+    return tmp;
+  }
+
+  /**
+   * @brief Addition assignment operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE friend T operator+(difference_type offset, T const& rhs)
+  {
+    T tmp(rhs);
+    tmp.p_ += (offset * rhs.width_);
+    return tmp;
+  }
+
+  /**
+   * @brief Compound assignment by difference operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE T& operator-=(difference_type offset)
+  {
+    T& derived = static_cast<T&>(*this);
+    derived.p_ -= offset * width_;
+    return derived;
+  }
+
+  /**
+   * @brief Decrement by offset operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE T operator-(difference_type offset) const
+  {
+    auto tmp = T(static_cast<T const&>(*this));
+    tmp.p_ -= (offset * width_);
+    return tmp;
+  }
+
+  /**
+   * @brief Subtraction assignment operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE friend T operator-(difference_type offset, T const& rhs)
+  {
+    T tmp(rhs);
+    tmp.p_ -= (offset * rhs.width_);
+    return tmp;
+  }
+
+  /**
+   * @brief Compute offset from iterator difference operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE difference_type operator-(T const& rhs) const
+  {
+    auto derived = static_cast<T const&>(*this);
+    return (derived.p_ - rhs.p_) / width_;
+  }
+
+  /**
+   * @brief Equals to operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE bool operator==(T const& rhs) const
+  {
+    return rhs.p_ == static_cast<T const&>(*this).p_;
+  }
+  /**
+   * @brief Not equals to operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE bool operator!=(T const& rhs) const
+  {
+    return rhs.p_ != static_cast<T const&>(*this).p_;
+  }
+  /**
+   * @brief Less than operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE bool operator<(T const& rhs) const
+  {
+    return static_cast<T const&>(*this).p_ < rhs.p_;
+  }
+  /**
+   * @brief Greater than operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE bool operator>(T const& rhs) const
+  {
+    return static_cast<T const&>(*this).p_ > rhs.p_;
+  }
+  /**
+   * @brief Less than or equals to operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE bool operator<=(T const& rhs) const
+  {
+    return static_cast<T const&>(*this).p_ <= rhs.p_;
+  }
+  /**
+   * @brief Greater than or equals to operator.
+   */
+  CUDA_HOST_DEVICE_CALLABLE bool operator>=(T const& rhs) const
+  {
+    return static_cast<T const&>(*this).p_ >= rhs.p_;
+  }
+
+ protected:
+  /**
+   * @brief Constructor assigns width and type member variables for base class.
+   */
+  base_indexalator(int width, data_type dtype) : width_(width), dtype_(dtype) {}
```

Comment:
Recommend using explicit `int32_t` here and two lines below (or `int16_t` if that works)
```suggestion
  base_indexalator(int32_t width, data_type dtype) : width_(width), dtype_(dtype) {}
```